### ISSUE TITLE
[jit] Support for union types

### ISSUE THREAD

**Issue by suo** [pytorch/pytorch#41412]
[PEP 484](https://www.python.org/dev/peps/pep-0484/) specifies Unions as part of the type system. We should consider supporting them in the TorchScript type system, as they are a common pattern in Python code. This is a low-pri request from PyPER as well.

**Comment by ppwwyyxx** [comment]
This would help detectron2 a lot as well. The concrete use cases on our side include:

1. Tensor-like structures where we'd like to define `__getitem__` as:
```
def __getitem__(self, idx: Union[int, slice, Tensor]):
```
because these types are all accepted by `torch.Tensor.__getitem__`. (slice is not supported but it's a different issue)

2. All our models return dict of losses in training, and prediction results in inference, i.e.
```
def forward(...) -> Union[Dict[str, Tensor], ResultType]
```

**Comment by shijianjian** [comment]
We are facing the same issue with Union param/return values.
```python
def forward(self, input: Union[Tensor, Tuple[Tensor, Tensor]]) -> Union[Tensor, Tuple[Tensor, Tensor]]:
```

@ansley @gmagogsfm Will this fix be included in the next version (0.7.0)?

**Comment by gmagogsfm** [comment]
> We are facing the same issue with Union param/return values.
> 
> ```python
> def forward(self, input: Union[Tensor, Tuple[Tensor, Tensor]]) -> Union[Tensor, Tuple[Tensor, Tensor]]:
> ```
> 
> @ansley @gmagogsfm Will this fix be included in the next version (0.7.0)?

Hi Shijianjian,

I assume you mean 1.7.0? If so, then fix will not be in it since 1.7 branch cut date has passed. But we will try to make it in 1.8.

**Comment by shijianjian** [comment]
Yes, 1.7.0. Sorry for the typo. Good to know that it will be supported in 1.8.0. Thank you!

**Comment by zh794390558** [comment]
so after 1.8.0, we can support below example?
```
def fun(bool cond) -> Union(Tuple[Tensor, Tensor], Tensor)
    if cond:
         return x, y
    else:
        return x
```

**Comment by ansley** [comment]
@zh794390558 Thanks for the question! Yes, we will support `Union` typed branch statements and `Union` with collection types in 1.8.0

**Comment by david-macleod** [comment]
Hi @ansley, is there any update on the progress of this? Or a link to another issue/MR I can track?

**Comment by shijianjian** [comment]
@gmagogsfm It seems still not available in the current nightly versions. Any updates on this?

**Comment by gmagogsfm** [comment]
Hi @shijianjian and @david-macleod,

We are still going through review process for this feature, sorry it is taking longer, we want to really get it right. Thanks for understanding.

**Comment by edgarriba** [comment]
@suo was this included in 1.8 release ?

**Comment by gmagogsfm** [comment]
> @suo was this included in 1.8 release ?

No, it has not landed yet.

**Comment by ProGamerGov** [comment]
The PyTorch documentation says that JIT support for Typing.Union is unlikely to be added: https://pytorch.org/docs/stable/jit_language_reference.html#unsupported-typing-constructs, but the comments in this PR imply that it is coming soon?

**Comment by gmagogsfm** [comment]
> The PyTorch documentation says that JIT support for Typing.Union is unlikely to be added: https://pytorch.org/docs/stable/jit_language_reference.html#unsupported-typing-constructs, but the comments in this PR imply that it is coming soon?

Yep, really soon thanks to @ansley

**Comment by shijianjian** [comment]
Good to know here. Is there any PR that we can track the process?

**Comment by sven-ds** [comment]
Thank you for implementing this - I'm right now building your branch! Did you guys already choose a target release for this feature?

**Comment by ansley** [comment]
@sven-ds We hope to have it out in 1.9! Please note that the current Union PR we have up has not been through a full code review. It should be fine to build the branch if you want to experiment, but there may be some small errors

**Comment by edgarriba** [comment]
keep the work up guys - union type is one of our missing pokemons to torchscript augmentations in kornia :)

**Comment by garymm** [comment]
Was this fixed by https://github.com/pytorch/pytorch/pull/64234? @ansley
